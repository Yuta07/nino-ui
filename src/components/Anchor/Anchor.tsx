import * as React from 'react';
import styled, { css } from 'styled-components';
import { Theme } from '../../themes/Theme';
import { useTheme } from '../../hooks/useTheme';

export type Props = {
  link: string;
  children?: React.ReactNode;
  size?: number;
  color?: string;
  width?: string;
  height: number;
  handleClick?: (event: React.MouseEvent<HTMLAnchorElement, MouseEvent>) => void;
};

export type AnchorProps = {
  iconName: string;
  iconSize?: number;
  iconColor?: string;
};

export const Anchor = ({ link, children, size, color = 'MAIN', width, height, handleClick }: Props) => {
  const theme = useTheme();

  return (
    <>
      <Base href={link} size={size} color={color} width={width} height={height} themes={theme} onClick={handleClick}>
        {children}
      </Base>
    </>
  );
};

const Base = styled.a<{
  size: Props['size'];
  color: Props['color'];
  width: Props['width'];
  height: Props['height'];
  themes: Theme;
}>`
  ${({ size, color, width, height, themes }) => {
    const { palette } = themes;

    return css`
      font-size: ${size}rem;
      color: ${palette.SECONDARY};
      background: ${palette[color]};
      width: ${width ? `${width}` : 'auto'};
      height: ${height}px;
      padding: 4px 0.5rem;
      display: inline-flex;
      justify-content: center;
      align-items: center;
      border: none;
      border-radius: 8px;
      text-align: center;
      text-decoration: none;
      cursor: pointer;
      white-space: nowrap;
      cursor: pointer;

      &:hover {
        opacity: 0.8;
        transition: 0.4s;
      }
    `;
  }}
`;
