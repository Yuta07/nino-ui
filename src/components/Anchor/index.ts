export * from './Anchor';
export * from './IconOnlyAnchor';
export * from './WithIconAnchor';
