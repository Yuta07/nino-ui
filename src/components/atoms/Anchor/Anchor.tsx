import * as React from 'react';
import styled, { css } from 'styled-components';
import { Theme } from '../../../themes/Theme';
import { useTheme } from '../../../hooks/useTheme';

export type Props = {
  link: string;
  children?: React.ReactNode;
  width?: string;
  height?: string;
  target?: string;
  handleClick?: (event: React.MouseEvent<HTMLAnchorElement, MouseEvent>) => void;
};

interface Interaction extends Props {
  border?: string;
}

export const Anchor = ({ link, children, width, height, target = '_self', handleClick }: Props) => {
  const theme = useTheme();

  return (
    <Base href={link} target={target} width={width} height={height} themes={theme} onClick={handleClick}>
      {children}
    </Base>
  );
};

export const UnderlineAnchor = ({ ...props }: Props) => {
  const theme = useTheme();

  return <AnchorWithUnderline themes={theme} onClick={props.handleClick} {...props} />;
};

export const InteractionAnchor = ({ border = 'DARK', ...props }: Interaction) => {
  const theme = useTheme();

  return <AnchorWithInteraction border={border} themes={theme} onClick={props.handleClick} {...props} />;
};

export const Base = styled.a<{
  width?: Props['width'];
  height?: Props['height'];
  themes: Theme;
}>`
  ${({ width, height, themes }) => {
    const { device, fontSize, palette } = themes;

    return css`
      width: ${width ? width : 'auto'};
      height: ${height ? height : 'auto'};
      font-size: ${fontSize.MEDIUM}px;
      color: ${palette.PRIMARY};
      display: inline-flex;
      align-items: center;
      justify-content: center;
      text-decoration: none;
      white-space: nowrap;
      cursor: pointer;

      &:hover {
        opacity: 0.8;
        transition: 0.2s;
      }

      @media screen and ${device.TABLET} {
        font-size: ${fontSize.SMALL}px;
      }

      @media screen and ${device.MOBILE_S} {
        font-size: ${fontSize.X_SMALL}px;
      }
    `;
  }}
`;

const AnchorWithUnderline = styled(Base)`
  &:hover {
    text-decoration: underline;
  }
`;

const AnchorWithInteraction = styled(Base)<{ border: Interaction['border']; themes: Theme }>`
  ${({ border, themes }) => {
    const { palette } = themes;

    return css`
      position: relative;

      &:hover {
        &:before {
          transform: scale3d(1, 1, 1);
          transform-origin: 0% 50%;
        }
      }

      &:before {
        content: '';
        color: ${palette[border]};
        position: absolute;
        top: 100%;
        left: 0;
        width: 100%;
        height: 2px;
        pointer-events: none;
        background: currentColor;
        transform: scale3d(0, 1, 1);
        transform-origin: 100% 50%;
        transition: transform 0.3s;
        transition-timing-function: cubic-bezier(0.8, 0, 0.2, 1);
      }

      @media (max-width: 425px) {
        padding: 20px 0 8px;
      }
    `;
  }}
`;
