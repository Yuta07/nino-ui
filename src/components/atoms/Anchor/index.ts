export * from './Anchor';
export * from './CustomAnchor';
export * from './IconOnlyAnchor';
export * from './WithIconAnchor';
