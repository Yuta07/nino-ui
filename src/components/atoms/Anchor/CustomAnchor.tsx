import * as React from 'react';
import styled, { css } from 'styled-components';
import { Theme } from '../../../themes/Theme';
import { useTheme } from '../../../hooks/useTheme';
import { Base, Props } from './Anchor';

export interface Custom extends Props {
  color?: string;
  background?: string;
}

export const CustomAnchor = ({ color = 'PRIMARY', background = 'SECONDARY', ...props }: Custom) => {
  const theme = useTheme();

  return <CustomAnchorStyle color={color} background={background} themes={theme} {...props} />;
};

const Border = css`
  border: none;
  border-radius: 8px;
`;

export const CustomAnchorStyle = styled(Base)<{
  color: Custom['color'];
  background: Custom['background'];
  themes: Theme;
}>`
  ${({ color, background, themes }) => {
    const { palette } = themes;

    return css`
      color: ${palette[color]};
      background: ${palette[background]};
      padding: 0.5rem 1rem;
      ${Border};
    `;
  }}
`;
