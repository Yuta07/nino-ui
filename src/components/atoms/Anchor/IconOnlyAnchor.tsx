import * as React from 'react';
import styled, { css } from 'styled-components';
import { Theme } from '../../../themes/Theme';
import { useTheme } from '../../../hooks/useTheme';
import { FeatherIcon } from '../Icon/FeatherIcon';
import { GithubIcon } from '../Icon/GithubIcon';
import { TypeIcon } from '../Icon/TypeIcon';
import { Base, Props } from './Anchor';
import { Custom } from './CustomAnchor';

type CustomProps = {
  iconName: string;
  iconSize?: number;
  iconColor?: string;
  border?: boolean;
  pattern: string;
  shape: 'square' | 'circle';
};

export type IconAnchorProps = Props & CustomProps & Custom;

export const IconOnlyAnchor = ({
  iconName,
  iconSize,
  iconColor,
  pattern,
  shape,
  border = false,
  color = 'PRIMARY',
  background = 'SECONDARY',
  ...props
}: IconAnchorProps) => {
  const theme = useTheme();

  let iconButton: JSX.Element;
  switch (pattern) {
    case 'Fi':
      iconButton = <FeatherIcon name={iconName} size={iconSize} color={theme.palette[iconColor]} />;
      break;
    case 'Go':
      iconButton = <GithubIcon name={iconName} size={iconSize} color={theme.palette[iconColor]} />;
      break;
    case 'Ti':
      iconButton = <TypeIcon name={iconName} size={iconSize} color={theme.palette[iconColor]} />;
      break;
    default:
      break;
  }

  return (
    <AnchorContainer color={color} background={background} border={border} shape={shape} themes={theme} {...props}>
      <IconButton>{iconButton}</IconButton>
    </AnchorContainer>
  );
};

const AnchorContainer = styled(Base)<{
  color: IconAnchorProps['color'];
  background: IconAnchorProps['background'];
  border: IconAnchorProps['border'];
  shape: IconAnchorProps['shape'];
  themes: Theme;
}>`
  ${({ color, background, border, shape, themes }) => {
    const { palette } = themes;

    return css`
      color: ${palette[color]};
      background: ${palette[background]};
      border: ${border ? `1px solid ${palette.DARK}` : 'none'};
      border-radius: ${shape === 'circle' ? '50%' : '8px'};
    `;
  }}
`;

const IconButton = styled.span`
  position: absolute;
  display: flex;
`;
