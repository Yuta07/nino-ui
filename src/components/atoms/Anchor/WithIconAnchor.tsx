import * as React from 'react';
import styled, { css } from 'styled-components';
import { Theme } from '../../../themes/Theme';
import { useTheme } from '../../../hooks/useTheme';
import { FeatherIcon } from '../Icon/FeatherIcon';
import { Base, Props } from './Anchor';
import { Custom } from './CustomAnchor';

type CustomProps = {
  iconName: string;
  iconSize?: number;
  iconColor?: string;
  position: string;
};

export type WithIconAnchorProps = Props & CustomProps & Custom;

export const WithIconAnchor = ({
  iconName,
  iconSize,
  iconColor,
  position,
  color = 'PRIMARY',
  background = 'SECONDARY',
  children,
  ...props
}: WithIconAnchorProps) => {
  const theme = useTheme();

  return (
    <>
      <AnchorContainer color={color} background={background} themes={theme} {...props}>
        {position === 'left' ? (
          <>
            <IconContainer>
              <FeatherIcon name={iconName} size={iconSize} color={theme.palette[iconColor]} />
            </IconContainer>
            <ButtonText position={position}>{children}</ButtonText>
          </>
        ) : (
          <>
            <ButtonText position={position}>{children}</ButtonText>
            <IconContainer>
              <FeatherIcon name={iconName} size={iconSize} color={theme.palette[iconColor]} />
            </IconContainer>
          </>
        )}
      </AnchorContainer>
    </>
  );
};

const Border = css`
  border: none;
  border-radius: 8px;
`;

const AnchorContainer = styled(Base)<{
  color: WithIconAnchorProps['color'];
  background: WithIconAnchorProps['background'];
  themes: Theme;
}>`
  ${({ color, background, themes }) => {
    const { device, fontSize, palette } = themes;

    return css`
      font-size: ${fontSize.MEDIUM}px;
      color: ${palette[color]};
      background: ${palette[background]};
      padding: 0.5rem 1rem;
      ${Border};

      &:hover {
        opacity: 0.8;
        transition: 0.4s;
      }

      @media screen and ${device.TABLET} {
        font-size: ${fontSize.SMALL}px;
      }

      @media screen and ${device.MOBILE_S} {
        font-size: ${fontSize.X_SMALL}px;
      }
    `;
  }}
`;

const IconContainer = styled.div`
  display: inline-flex;
`;

const ButtonText = styled.span<{ position: WithIconAnchorProps['position'] }>`
  ${({ position }) => {
    return css`
      ${position === 'left' ? `margin-left: 12px;` : `margin-right: 12px;`}
    `;
  }}
`;
