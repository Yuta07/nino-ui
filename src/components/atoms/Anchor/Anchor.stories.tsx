import { storiesOf } from '@storybook/react';
import * as React from 'react';
import styled from 'styled-components';
import { Anchor, UnderlineAnchor, InteractionAnchor } from './Anchor';
import { CustomAnchor } from './CustomAnchor';
import { IconOnlyAnchor } from './IconOnlyAnchor';
import { WithIconAnchor } from './WithIconAnchor';

const Wrapper = () => {
  return (
    <>
      <Container>
        <Anchor link="#" height="32px">
          Anchor - Original
        </Anchor>
      </Container>
      <Container>
        <UnderlineAnchor link="#" height="32px">
          Anchor - Underline
        </UnderlineAnchor>
      </Container>
      <Container>
        <InteractionAnchor border="MAIN" link="#" height="32px">
          Anchor - Interaction
        </InteractionAnchor>
      </Container>
      <Container>
        <CustomAnchor link="#" height="32px">
          Custom - Normal
        </CustomAnchor>
      </Container>
      <Container>
        <CustomAnchor link="#" height="32px" color="SECONDARY" background="MAIN">
          Custom - background: MAIN, color: SECONDARY
        </CustomAnchor>
      </Container>
      <Container>
        <WithIconAnchor
          link="#"
          iconName="Fi-Heart"
          iconSize={16}
          iconColor="SECONDARY"
          width="100%"
          color="SECONDARY"
          background="DANGER"
          position="left"
        >
          Favorite
        </WithIconAnchor>
      </Container>
      <Container>
        <WithIconAnchor
          link="#"
          iconName="Fi-Share2"
          iconColor="SECONDARY"
          width="100%"
          color="SECONDARY"
          background="INFO"
          position="Right"
        >
          Share
        </WithIconAnchor>
      </Container>
      <IconOnlyContainer>
        <IconOnlyAnchor
          iconName="Go-Zap"
          iconSize={16}
          iconColor="SECONDARY"
          link=""
          width="32px"
          height="32px"
          background="WARNING"
          pattern="Go"
          shape="circle"
        />
        <IconOnlyAnchor
          iconName="Go-Bell"
          iconSize={16}
          iconColor="SECONDARY"
          link=""
          width="32px"
          height="32px"
          background="INFO"
          pattern="Go"
          shape="circle"
        />
        <IconOnlyAnchor
          iconName="Ti-Cog"
          iconSize={16}
          iconColor="PRIMARY"
          link=""
          width="32px"
          height="32px"
          background="GRAY"
          pattern="Ti"
          shape="square"
        />
        <IconOnlyAnchor
          iconName="Ti-WeatherStormy"
          iconSize={16}
          iconColor="SECONDARY"
          link=""
          width="32px"
          height="32px"
          background="SUCCESS"
          pattern="Ti"
          shape="circle"
        />
      </IconOnlyContainer>
    </>
  );
};

storiesOf('Anchor', module).add('all', () =>
  React.createElement(() => {
    return <Wrapper />;
  })
);

const Container = styled.div`
  margin: 2rem 0;
`;

const IconOnlyContainer = styled.div`
  display: flex;
  flex-direction: row;
  justify-content: space-around;
  align-items: center;
  width: 100%;
  margin: 2rem 0;
`;
