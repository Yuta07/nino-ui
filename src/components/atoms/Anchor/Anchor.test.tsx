import React from 'react';
import * as renderer from 'react-test-renderer';

import { Anchor } from './Anchor';

describe('Anchor test', function() {
  test('Nothing changes when clicked', () => {
    const component = renderer.create(
      <Anchor link="#" width="120px" height="32px">
        nino-ui
      </Anchor>
    );
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });
});
