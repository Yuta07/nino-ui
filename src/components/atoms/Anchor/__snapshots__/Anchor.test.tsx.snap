// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Anchor test Nothing changes when clicked 1`] = `
.c0 {
  font-size: 1rem;
  color: #fefefe;
  background: #f39c12;
  width: 120px;
  height: 32px;
  padding: 4px 0.5rem;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  border: none;
  border-radius: 8px;
  text-align: center;
  -webkit-text-decoration: none;
  text-decoration: none;
  cursor: pointer;
  white-space: nowrap;
}

.c0:hover {
  opacity: 0.8;
  -webkit-transition: 0.4s;
  transition: 0.4s;
}

@media screen and (max-width:768px) {
  .c0 {
    font-size: 14px;
  }
}

@media screen and (max-width:320px) {
  .c0 {
    font-size: 12px;
  }
}

<a
  className="c0"
  color="MAIN"
  height={32}
  href="#"
  size={1}
  width="120px"
>
  nino-ui
</a>
`;
